lib/jack_tokenizer.rb

```diff
@@ -1,34 +1,47 @@
 class JackTokenizer
   def initialize(io)
-    @raw_tokens = io.read.split(/\s(?=(?:[^"]|"[^"]*")*$)/)
+    @input = io.read
+    @index = 0
   end
 
   def has_more_tokens?
-    !@raw_tokens.empty?
+    @index < @input.length
   end
 
   def advance
-    @current_token = @raw_tokens.shift
+    if @input[@index] == '"'
+      next_index = @input.index('"', @index + 1) + 1
```

Comment:
This code is good for right now, but bear in mind that at this point we _already know_ that the token type is going to be `:STRING_CONST`, because we’re making that assumption when we go looking for another `"` as the end of the token. So it’s a little silly that we instantly forget that information and then have a whole separate conditional later in the method for determining the token type — in this specific case, at least, we already knew it.

It’s not a big deal at the moment, but if this logic gets more complex it might be beneficial for us to consolidate the “find the end of the current token“ and “determine the type of the current token” pieces which are currently separate. Modularity is nice to have, but only if the modular pieces are genuinely independent; in this case I have a gut feeling that “where does the token end?” and “what type of token is it?” are actually not independent questions (right?), so trying to answer them separately might make our implementation do more work than is necessary.